### PR TITLE
Open-source info updates

### PR DESCRIPTION
**Pull request by ameza13**
---
## Status
**READY**

## Description
Integrate CoC, extensions outline, contribution guide updates for open-source, added a github template for documentation.

## Impacted Areas in Library
NA

## Which issue(s) does this pull-request fix?
Closes: #30 

## Any special notes for your reviewer?

These are the files added/updated:

- Development guide (updated)
- Information on open-source extensions (new)
- Code of conduct (new)
- Main README file includes links to the new resources (updated)
- Added a template for documentation to `.github`

---

## Checklist
- [ ] Automated tests exist
- [ ] Updated Package Requirements (if required, and with maintainers' approval)
- [ ] Local unit tests performed
- [X] Documentation exists [link]()
- [X] Local pre-commit hooks performed
- [X] Desired commit message set as PR title and description set above
- [X] Link to relevant GitHub issue provided

### CODE REVIEW

File: EXTENSIONS.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+- [Design considerations](#design-considerations-to-build-extensions)
+
+# IBM Generative AI Core Package
+[IBM Generative AI open-source repository](https://github.com/IBM/ibm-generative-ai) lives under [IBM Github organization](https://github.com/IBM/). This repository is the official and unique location of IBM Gen AI core package source code.
+
+# What is an IBM Generative AI Extension?
+An extension is a software add-on that is installed on a program to enhance its capabilities. Gen AI extensions are meant to expand the capabilities of the Gen AI core package. For instance, imagine that you want to build a prompt pattern with data from a pandas dataframe. You could write your own code  to achive such functionaliy, or you could instead use the Gen AI Pandas extension. Using the extension will allow you to save development time and deliver an elegant solution that re-uses code when it is possible.
+
+# IBM Generative AI Extension types
+IBM Generative AI supports three two types of open-source extensions: open-source official, and open-source third-party extensions.
```
Comment (onkarbhardwaj) [inline EXTENSIONS.md:18]:
```suggestion
IBM Generative AI supports two types of open-source extensions: open-source official, and open-source third-party extensions.
```

File: EXTENSIONS.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+## Open-source "Gen AI official" extensions
+Extensions that are meant for public use from get-go should instead be developed as open-source extensions. Examples of open-source extensions that have been already released as open-source are LangChain and Hugging Face extensions.
```
Comment (onkarbhardwaj) [inline EXTENSIONS.md:21]:
```suggestion
Extensions that are meant for public use from get-go should instead be developed as open-source extensions. Examples of open-source extensions that have been already released as open-source are LangChain, Pandas, and Hugging Face extensions.
```

File: EXTENSIONS.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+Extensions might be migrated from one type to another upon evaluation and approval of the Gen AI maintenance team. Below, we describe three common migration scenarios:
+
+## Open-source "third-party" to open-source "Gen AI official"
+If open-source third-party contributors have an interest in their extensions becoming official, they must contact the IBM Gen AI team. The Gen AI team will evaluate the request considering novelty and implementation quality (meaning it was developed following the Gen AI official guide and meets Gen AI extensions' design considerations). If the request is approved, the Gen AI team will assist the extension owners in migrating the extension.
+
+## Open-source "Gen AI official" to Gen AI's core
+Upon evaluation of the Gen AI maintenance team, open-source extensions could be merged with Gen AI core code and become deprecated as extensions. Deprecation announcements will be shared as part of the release notes of every new version of the IBM Gen AI package.
+
+# Design considerations to build extensions
+To design an extension for IBM Gen AI SDK a developer should consider a mix-and-match of the following two important development experience patterns:
```
Comment (onkarbhardwaj) [inline EXTENSIONS.md:44]:
```suggestion
To design an extension for IBM Gen AI SDK a developer should consider a mix-and-match of the following two important development experience patterns among other patterns:
```

File: EXTENSIONS.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+try:
+    from langchain.llms.base import LLM
+    from langchain.llms.utils import enforce_stop_tokens
+except ImportError:
+    raise ImportError("Could not import langchain: Please install ibm-generative-ai[langchain] extension.")
+
+...
+
+class LangChainInterface(LLM, BaseModel):
```
Comment (onkarbhardwaj) [inline EXTENSIONS.md:71]: Maybe need to add some lines below so that it does not feel abruptly cut off.

File: EXTENSIONS.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+The code snippet below shows an example of this pattern. While `from_template` method is part of an extension class (`PromptExtension`) and not a core class (`PromptPattern`), the way it is being called gives end-users the impression the method is part of `PromptPattern` core class. The magic that enables providing this development experience is the use of decorators at import time. This pattern enables extending user-facing classes in a user-friendly way.
+
+`genai_prompt_pattern = PromptPattern.langchain.from_template(langchain_prompt_template)  # notice langchain hook in the function call.`
+
+## Extensions that wrap core class functionality
+
+Designing extensions that do not integrate, but  **wrap functionality in Gen AI's core** for common use cases. These classes are meant to be used as entities on their own.
+
+The `LangChainInterface` class (see below) is an example of this pattern. `LangChainInterface` is a subclass of langchain's `LLM` class that wraps `generate` functionality of Gen AI's Model.
```
Comment (onkarbhardwaj) [inline EXTENSIONS.md:58]: Maybe end this description something like "For full definition of this class, see <link-from-langchain-file-main-branch>"?

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -40,13 +41,20 @@ source .venv/bin/activate # Activate and use your virtual environment
 # And once you're all done..
 deactivate
 ```
+### <a name='Forkrepo'></a>Fork the official repo
```
Comment (onkarbhardwaj) [inline DEVELOPMENT.md:44]: I think these links likely generated with vscode extensions don't work with github. Would it be possible for you to correct the links in this document(not necessarily the result of your contributions) based on the format in README?

```markdown
* [Table of Contents](#table-of-contents)
* [Installation](#installation)
* [Gen AI Endpoint](#gen-ai-endpoint)
    * [Example](#example)
* [Examples](#examples)
    * [Async Example](#async-example)
    * [Synchronous Example](#synchronous-example)
* [Tips and Troubleshooting](#tips-and-troubleshooting)
```

File: EXTENSIONS.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+```
```
Comment (nzw0301) [inline EXTENSIONS.md:60]:
```suggestion
```python
```

File: EXTENSIONS.md (hunk 1 above)
Comment (ameza13) [reply EXTENSIONS.md]: While updating the links, I noticed the three last sections do not exist in this file:

```
* [Deployment Process](#DeploymentProcess)
	* [Automated Release Process](#AutomatedReleaseProcess)
	* [Manual Release Process](#ManualProcess)

```

Should I remove them temporarily? do we have this info somewhere?

File: DEVELOPMENT.md (hunk 6 above)
Comment (ameza13) [reply DEVELOPMENT.md]: While updating the index for DEVELOP.md, I noticed the three last sections do not exist in this file:

```
* [Deployment Process](#DeploymentProcess)
	* [Automated Release Process](#AutomatedReleaseProcess)
	* [Manual Release Process](#ManualProcess)

```

Temporarily, I commented them. Do we have this info somewhere else?

**Review by nzw0301 — APPROVED** [review]
LGTM on my suggestion part. Thanks!

File: ACTIVE_PROJECT_TEAM.md [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# IBM Generative IA Active Maintainers
```
Comment (vedem1192) [inline ACTIVE_PROJECT_TEAM.md:1]: Small typo:
```suggestion
# IBM Generative AI Active Maintainers
```

File: EXTENSIONS.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+# What is an IBM Generative AI Extension?
+An extension is a software add-on that is installed on a program to enhance its capabilities. Gen AI extensions are meant to expand the capabilities of the Gen AI core package. For instance, imagine that you want to build a prompt pattern with data from a pandas dataframe. You could write your own code  to achive such functionaliy, or you could instead use the Gen AI Pandas extension. Using the extension will allow you to save development time and deliver an elegant solution that re-uses code when it is possible.
+
+# IBM Generative AI Extension types
+IBM Generative AI supports two types of open-source extensions: open-source official, and open-source third-party extensions.
+
+## Open-source "Gen AI official" extensions
+Extensions that are meant for public use from get-go should instead be developed as open-source extensions. Examples of open-source extensions that have been already released as open-source are LangChain, Pandas, and Hugging Face extensions.
```
Comment (vedem1192) [inline EXTENSIONS.md:21]: Is this a better way of separating supported open-source extension and 3rd-party ones?
```suggestion
Extensions that are meant for public use from the get-go should instead be developed as official open-source extensions. Examples of official extensions that have already been released are LangChain, Pandas, and Hugging Face extensions.
```

File: EXTENSIONS.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+### Ownership and location
+Open-source extensions should be submitted directly to the [IBM Generative AI open-source repository](https://github.com/IBM/ibm-generative-ai) and should be developed following the [open-source Gen AI contribution guide](https://github.com/IBM/ibm-generative-ai/blob/main/DEVELOPMENT.md). Open-source official extensions are typically developed by the Gen AI team, or in collaboration with them. Providing maintenance to open-source official extensions is responsability of the Gen AI team.
+
+## Open-source "third-party" extensions
+We refer to all other extensions not implemented nor officially maintained by the Gen AI team as open-source third-party extensions.
```
Comment (vedem1192) [inline EXTENSIONS.md:27]:
```suggestion
All other extensions neither implemented nor officially maintained by the Gen AI team are referred to as open-source third-party extensions.
```

File: EXTENSIONS.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+### Ownership and location
+Open-source third-party extensions can live anywhere, inside [IBM organization](https://github.com/IBM/), the wider Github universe, or other contributors' own Github spaces (e.g., other IBM teams, clients, partners). Maintaining these extensions would be the responsibility of their owners.
+
+# Extensions migration
+Extensions might be migrated from one type to another upon evaluation and approval of the Gen AI maintenance team. Below, we describe three common migration scenarios:
```
Comment (vedem1192) [inline EXTENSIONS.md:33]: I think there's only 2 scenario at the moment.
```suggestion
Extensions might be migrated from one type to another upon evaluation and approval of the Gen AI maintenance team. Below, we describe two common migration scenarios:
```

File: EXTENSIONS.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+Upon evaluation of the Gen AI maintenance team, open-source extensions could be merged with Gen AI core code and become deprecated as extensions. Deprecation announcements will be shared as part of the release notes of every new version of the IBM Gen AI package.
+
+# Design considerations to build extensions
+Among other patterns, a developer should consider a mix-and-match of the two following important development experience patterns to design an extension for IBM Gen AI SDK:
+
+## Extensions that simulate being part of a core class
+
+In spite of the extension's code living in a separate location, **try to give the impression that extended functionality is part of an IBM Gen AI core class**.
+
+The code snippet below shows an example of this pattern. While `from_template` method is part of an extension class (`PromptExtension`) and not a core class (`PromptPattern`), the way it is being called gives end-users the impression the method is part of `PromptPattern` core class. The magic that enables providing this development experience is the use of decorators at import time. This pattern enables extending user-facing classes in a user-friendly way.
```
Comment (vedem1192) [inline EXTENSIONS.md:48]:
```suggestion
The code snippet below shows an example of this pattern. While `from_template` method is part of an extension class (`PromptExtension`) and not a core class (`PromptPattern`), the way it is being called gives end-users the impression that the method is part of the `PromptPattern` core class. The magic that enables providing this development experience is the use of decorators at import time. This pattern enables extending user-facing classes in a user-friendly way.
```

File: EXTENSIONS.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+- [Design considerations](#design-considerations-to-build-extensions)
+
+# IBM Generative AI Core Package
+[IBM Generative AI open-source repository](https://github.com/IBM/ibm-generative-ai) lives under [IBM Github organization](https://github.com/IBM/). This repository is the official and unique location of IBM Gen AI core package source code.
+
+# What is an IBM Generative AI Extension?
+An extension is a software add-on that is installed on a program to enhance its capabilities. Gen AI extensions are meant to expand the capabilities of the Gen AI core package. For instance, imagine that you want to build a prompt pattern with data from a pandas dataframe. You could write your own code  to achive such functionaliy, or you could instead use the Gen AI Pandas extension. Using the extension will allow you to save development time and deliver an elegant solution that re-uses code when it is possible.
+
+# IBM Generative AI Extension types
+IBM Generative AI extensions can be either of the following :
```
Comment (vedem1192) [inline EXTENSIONS.md:18]:
```suggestion
IBM Generative AI extensions can be either of the following :
```

**Review by onkarbhardwaj — APPROVED** [review]

File: EXTENSIONS.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+- [Design considerations](#design-considerations-to-build-extensions)
+
+# IBM Generative AI Core Package
+[IBM Generative AI open-source repository](https://github.com/IBM/ibm-generative-ai) lives under [IBM Github organization](https://github.com/IBM/). This repository is the official and unique location of IBM Gen AI core package source code.
+
+# What is an IBM Generative AI Extension?
+An extension is a software add-on that is installed on a program to enhance its capabilities. Gen AI extensions are meant to expand the capabilities of the Gen AI core package. For instance, imagine that you want to build a prompt pattern with data from a pandas dataframe. You could write your own code  to achive such functionaliy, or you could instead use the Gen AI Pandas extension. Using the extension will allow you to save development time and deliver an elegant solution that re-uses code when it is possible.
+
+# IBM Generative AI Extension types
+IBM Generative AI supports two types of open-source extensions: open-source official, and open-source third-party extensions.
```
[the remark this answers is not in the document]
Comment (ameza13) [reply EXTENSIONS.md]: rephrase to something very close, but shorter.

IBM Generative AI open-source extensions can be either **official** or **third-party** extensions.

File: EXTENSIONS.md (hunk 9 above)
Comment (ameza13) [reply EXTENSIONS.md]: Thanks, I agree.

File: EXTENSIONS.md (hunk 11 above)
Comment (ameza13) [reply EXTENSIONS.md]: right, good catch!

**Pull request merged**